printing/printer.ml

```diff
@@ -332,15 +338,44 @@ let pr_ne_context_of header env sigma =
     List.is_empty (Environ.named_context env)  then (mt ())
   else let penv = pr_context_unlimited env sigma in (header ++ penv ++ fnl ())
 
+(* Heuristic for horizontalizing hypothesis:
+   Detecting variable which type is a simple id *)
+let should_compact env sigma typ =
+  is_compact_context() &&
+  match kind_of_term typ with
+  | Rel _ | Var _ | Sort _ | Const _ | Ind _ -> true
+  | _ ->
+     let _,type_of_typ = Typing.type_of env sigma typ in
+     match kind_of_term type_of_typ with
+     | Sort s -> not (Sorts.is_prop s)
+     | _ -> true
+
+(* If option Compact Contexts is set, we pack "simple" hypothesis in a
+   hov box (with three sapaces as a separator), the global box being a
+   v box *)
 let rec bld_sign_env env sigma ctxt pps =
   match ctxt with
   | [] -> pps
+  | CompactedDecl.LocalAssum (ids,typ)::ctxt' when should_compact env sigma typ ->
+    let pps',ctxt' = bld_sign_env_id env sigma ctxt (mt ()) true in
+    (* putting simple hyps in a more horizontal flavor *)
+    bld_sign_env env sigma ctxt' (pps ++ brk (0,0) ++ hov 0 pps')
   | d:: ctxt' ->
-     let pidt = pr_var_list_decl env sigma d in
-     let pps' = pps ++ brk (0,0) ++ pidt in
-     bld_sign_env env sigma ctxt' pps'
+    let pidt = pr_var_list_decl env sigma d in
+    let pps' = pps ++ brk (0,0) ++ pidt in
+    bld_sign_env env sigma ctxt' pps'
+and bld_sign_env_id env sigma ctxt pps is_start =
+  match ctxt with
+  | [] -> pps,ctxt
+  | CompactedDecl.LocalAssum(ids,typ) as d :: ctxt' when should_compact env sigma typ ->
+    let pidt = pr_var_list_decl env sigma d in
+    let pps' = pps ++ (if not is_start then brk (3,0) else (mt ())) ++ pidt in
```

Comment:
I don't understand the rationale for picking three spaces as a separator. Since the line already contains some colons and commas, it seems like it would be more readable to choose a stronger delimiter than just spaces. Why not a semicolon for example? It also means that we could later allow slightly less simple types, e.g. `list nat`, if needs arise.

Also, no need for the extra parentheses around `mt ()`.